components/connectionmanager/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+        let manager = Arc::new(Self {
+            p2p_adaptor,
+            outbound_target,
+            inbound_limit,
+            amgr,
+            connection_requests: Default::default(),
+            force_next_iteration: tx,
+            shutdown_signal: shutdown_signal_tx,
+        });
+        manager.clone().start_event_loop(rx, shutdown_signal_rx);
+        manager.force_next_iteration.send(()).unwrap();
+        manager
+    }
+
+    fn start_event_loop(self: Arc<Self>, mut rx: UnboundedReceiver<()>, mut shutdown_signal_rx: UnboundedReceiver<()>) {
+        let mut ticker = interval(Duration::from_secs(30));
+        tokio::spawn(async move {
+            loop {
+                select! {
+                    _ = rx.recv() => self.clone().handle_event().await,
+                    _ = ticker.tick() => self.clone().handle_event().await,
+                    _ = shutdown_signal_rx.recv() => break,
+                }
+            }
+            debug!("Connection manager event loop exiting");
+        });
+    }
+
+    async fn handle_event(self: Arc<Self>) {
+        debug!("Starting connection loop iteration");
+        let peers = self.p2p_adaptor.active_peers();
+        let peer_by_address: HashMap<SocketAddr, Peer> = peers.into_iter().map(|peer| (peer.net_address(), peer)).collect();
+
+        self.handle_connection_requests(&peer_by_address).await;
+        self.handle_outbound_connections(&peer_by_address).await;
+        self.handle_inbound_connections(&peer_by_address).await;
+    }
+
+    pub async fn add_connection_request(&self, address: SocketAddr, is_permanent: bool) {
+        // If the request already exists, it resets the attempts count and overrides the `is_permanent` setting.
+        self.connection_requests
+            .lock()
+            .await
+            .insert(address, ConnectionRequest { next_attempt: SystemTime::now(), is_permanent, attempts: 0 });
+        self.force_next_iteration.send(()).unwrap(); // We force the next iteration of the connection loop.
+    }
+
+    pub async fn stop(&self) {
+        self.shutdown_signal.send(()).unwrap();
+    }
+
+    async fn handle_connection_requests(self: &Arc<Self>, peer_by_address: &HashMap<SocketAddr, Peer>) {
+        let mut requests = self.connection_requests.lock().await;
+        let mut new_requests = HashMap::with_capacity(requests.len());
+        for (address, request) in requests.iter() {
+            let address = *address;
+            let request = request.clone();
+            let is_connected = peer_by_address.contains_key(&address);
+            if is_connected && !request.is_permanent {
+                continue;
+            }
+
+            if !is_connected && request.next_attempt <= SystemTime::now() {
+                debug!("Connecting to a connection request to {}", address);
+                if self.p2p_adaptor.connect_peer(address.to_string()).await.is_none() {
```

Comment:
Is this `connect_peer` (with the default 16 retries) deliberate? I suggest making the default w/o retries, and rename `connect_peer_with_retry_params` -> `connect_peer_with_retries` and fix current test locations that call `connect_peer` to use the latter